src/mcpwm.rs

```diff
@@ -0,0 +1,597 @@
+        self
+    }
+
+    #[must_use]
+    pub fn deadtime(mut self, deadtime: impl Into<Option<DeadtimeConfig>>) -> Self {
+        self.deadtime = deadtime.into();
+        self
+    }
+}
+
+impl Default for OperatorConfig {
+    fn default() -> Self {
+        Self {
+            frequency: 1000.Hz(),
+            duty_a: 50.0,
+            duty_b: 50.0,
+
+            #[cfg(not(esp_idf_version = "4.3"))]
+            lowest_frequency: 16.Hz(),
+
+            duty_mode: DutyMode::ActiveHigh,
+            counter_mode: CounterMode::Up,
+
+            deadtime: None,
+        }
+    }
+}
+
+#[derive(Default)]
+pub struct UnitZero;
+
+#[derive(Default)]
+pub struct UnitOne;
+
+pub type Duty = f32;
+
+pub struct MCPWM<U: Unit> {
+    _unit: U,
+}
+
+impl<U: Unit> MCPWM<U> {
+    /// # Safety
+    ///
+    /// It is safe to instantiate this exactly one time per `Unit`.
+    unsafe fn new() -> Self {
+        Self {
+            _unit: U::default(),
+        }
+    }
+}
+
+pub trait Unit: Default {
+    fn unit() -> mcpwm_unit_t;
+}
+
+impl Unit for UnitZero {
+    fn unit() -> mcpwm_unit_t {
+        mcpwm_unit_t_MCPWM_UNIT_0
+    }
+}
+
+impl Unit for UnitOne {
+    fn unit() -> mcpwm_unit_t {
+        mcpwm_unit_t_MCPWM_UNIT_1
+    }
+}
+
+// TODO: How do we want fault module to fit into this?
+
+pub struct Mcpwm<U: Unit> {
+    #[cfg(not(esp_idf_version = "4.3"))]
+    /// This is the frequency of the clock signal passed on as clock source for the operators timers
+    /// Those timers in turn have their own prescalers to scale this down even further
+    ///
+    /// NOTE: This is only to be set by calling Self::lowest_frequency
+    operator_input_frequency: u32,
+    _instance: MCPWM<U>,
+}
+
+impl<U: Unit> Mcpwm<U> {
+    pub fn new(instance: MCPWM<U>) -> Result<Self, EspError> {
+        let res = Self {
+            #[cfg(not(esp_idf_version = "4.3"))]
+            operator_input_frequency: 0,
+            _instance: instance,
+        };
+
+        #[cfg(not(esp_idf_version = "4.3"))]
+        {
+            res.lowest_frequency(15.Hz())
+        }
+
+        #[cfg(esp_idf_version = "4.3")]
+        {
+            Ok(res)
+        }
+    }
+
+    /// Specify lowest reachable frequency
+    ///
+    /// The lower this is set, the lower frequencies will be reachable. However, this is at the cost of worse
+    /// resolution at higher frequencies.
+    ///  
+    /// Same thing goes for the other way. The higher value set here, the more resolution and so on.
+    #[cfg(not(esp_idf_version = "4.3"))]
+    pub fn lowest_frequency(mut self, lowest_frequency: Hertz) -> Result<Self, EspError> {
+        // TODO: Do we care about frequency < 1Hz?
+        let operator_input_frequency =
+            MAX_PWM_TIMER_PRESCALE * MAX_PWM_TIMER_PERIOD * u32::from(lowest_frequency);
+        let group_pre_scale = MCPWM_CLOCK_SOURCE_FREQUENCY / operator_input_frequency;
+        if !(1..=256).contains(&group_pre_scale) {
```

Comment:
Doing some quick math:

```
let operator_input_frequency =
            MAX_PWM_TIMER_PRESCALE * MAX_PWM_TIMER_PERIOD * u32::from(lowest_frequency);
let group_pre_scale = MCPWM_CLOCK_SOURCE_FREQUENCY / operator_input_frequency;
<=>

let group_pre_scale = 160MHz / (2^8 * 2^16 * lowest_frequency);
```
With smallest nonzero frequency value
```
lowest_frequency = 1 => largest achievable group_pre_scale ~= 9.54 ~= 9
```
A `lowest_frequency` as small as 5Hz would lead to a `group_pre_scale ~= 1.9 ~= 1`

so our usable range of `lowest_frequency` is 1Hz to 5Hz inclusive. Which feels, not great...

----

In my mind we have some options:

1. Go with what the IDF does: expose `group_resolution`. This value is(if I understand correctly) the frequency of the clock signal that the underlaying operators timers will have as their input. (they can then apply both an 8-bit prescaler and a 16-bit period to that for the final PWM signal)
2. Expose `group_prescaler`. This is closest to what the hardware actually does. `group_prescaler` is the factor by which the 160MHz clock signal will be scaled down before it is passed on to the underlaying operators timers.
3. Stick with the current proposal `lowest_frequency` which is the lowest frequency the operators timer will be able to reach. Consider using some other frequency type to be able to reach frequencies < 1Hz
4. Cobinations of the above. We could have multiple methods all leading to the same end result depending on the users preference.
5. Other suggestions

Reply:
Also should MCPWM be marked as extra experimental or something until we are more sure of how it should look? If so, how?

Reply:
Would it be possible to work backwards towards a target frequency instead? I know some other HALs allow target frequencies to be set, then the initialization takes care of setting the various prescalers or selecting different clock sources, and obviously throws an error if the clock is not reachable. Haven't looked into the feasibility of this, but it might be an option.

The above suggestion might be a far whack of work, so the simplest `lowest_frequency` approach is definitely welcome on the first run.

Reply:
Not quite sure what you mean by
> Would it be possible to work backwards towards a target frequency instead?

If you mean that for example something like calling `Operator::set_frequency()` would make sure that frequency is reachable and automagically adjust the group prescaler in such a way as to maximize the resolution for all three operators?

I personally think that would be a little bit too far. That would for example lead to things like dead times being affected and, I am not sure, but possibly all sort of other things like carrier and capture modules. All due to a simple call to `Operator::set_frequency`.

That is also why I (atleast for now) made `Mcpwm::lowest_frequency` impossible to call while having existing `Operator`s from that mcpwm module.

Anyways on the other hand, perhaps making `Operator::lowest_frequency` (the individual timer's prescaler, not group prescaler) could be made in such a was as to be automagically set by `Operator::set_frequency`?